cmd/profilecli/blocks_tool.go

```diff
@@ -0,0 +1,365 @@
+	blockGroupMap := make(map[string]*blockGroup)
+	blockGroups := make([]*blockGroup, 0)
+
+	deletedBlocks := make(map[ulid.ULID]int64)
+	if !query.IncludeDeleted {
+		for _, deletionMark := range index.BlockDeletionMarks {
+			deletedBlocks[deletionMark.ID] = deletionMark.DeletionTime
+		}
+	}
+
+	for _, blk := range index.Blocks {
+		if _, deleted := deletedBlocks[blk.ID]; !deleted && blk.Within(queryFrom, queryTo) {
+			minTime := blk.MinTime.Time().UTC().Format(time.RFC3339)
+			blkGroup, ok := blockGroupMap[minTime]
+			if !ok {
+				blkGroup = &blockGroup{
+					MinTime:    minTime,
+					Blocks:     make([]*blockDetails, 0),
+					MinTimeAge: humanize.RelTime(blk.MinTime.Time(), time.Now(), "ago", ""),
+				}
+				blockGroups = append(blockGroups, blkGroup)
+			}
+			blkGroup.Blocks = append(blkGroup.Blocks, &blockDetails{
+				ID:               blk.ID.String(),
+				MinTime:          minTime,
+				MaxTime:          blk.MaxTime.Time().UTC().Format(time.RFC3339),
+				Duration:         blk.MaxTime.Sub(blk.MinTime).String(),
+				UploadedAt:       blk.GetUploadedAt().UTC().Format(time.RFC3339),
+				CompactorShardID: blk.CompactorShardID,
+			})
+			blockGroupMap[minTime] = blkGroup
+		}
+	}
+
+	slices.SortFunc(blockGroups, func(a, b *blockGroup) bool {
+		return a.MinTime > b.MinTime
+	})
+
+	return postProcessBlockGroups(blockGroups)
+}
+
+func postProcessBlockGroups(blockGroups []*blockGroup) []*blockGroup {
+	for i := 0; i < len(blockGroups)-1; i += 1 {
+		blockGroup := blockGroups[i]
+		if !strings.Contains(blockGroup.MinTime, "0:00Z") {
+			nextGroup := blockGroups[i+1]
+			nextGroup.Blocks = append(nextGroup.Blocks, blockGroup.Blocks...)
+			blockGroup.Blocks = make([]*blockDetails, 0)
+		}
+		slices.SortFunc(blockGroup.Blocks, func(a, b *blockDetails) bool {
+			return a.MinTime > b.MinTime
+		})
+	}
+
+	finalBlockGroups := make([]*blockGroup, 0)
+	for _, blockGroup := range blockGroups {
+		if len(blockGroup.Blocks) > 0 {
+			finalBlockGroups = append(finalBlockGroups, blockGroup)
+		}
+	}
+	return finalBlockGroups
+}
+
+func getBlockDetails(ctx context.Context, id ulid.ULID, fetcher *block.MetaFetcher) *blockDetails {
+	meta, err := fetcher.LoadMeta(ctx, id)
+	if err != nil {
+		return nil
+	}
+	var blockSize uint64
+	for _, f := range meta.Files {
+		blockSize += f.SizeBytes
+	}
+
+	return &blockDetails{
+		ID:              meta.ULID.String(),
+		MinTime:         meta.MinTime.Time().UTC().Format(time.RFC3339),
+		MaxTime:         meta.MaxTime.Time().UTC().Format(time.RFC3339),
+		Duration:        meta.MaxTime.Sub(meta.MinTime).String(),
+		CompactionLevel: meta.Compaction.Level,
+		Size:            humanize.Bytes(blockSize),
+		Stats:           meta.Stats,
+		Labels:          meta.Labels,
+	}
+}
+
+func (t *blocksWebTool) run(ctx context.Context) error {
+	out := output(ctx)
+
+	fmt.Fprintf(out, "The blocks web tool is available at http://localhost:%d/blocks/index\n", t.params.httpListenPort)
+
```

Comment:
Nice to have: Potentially open the browser automatically: https://stackoverflow.com/questions/39320371/how-start-web-server-to-open-page-in-browser-in-golang

Reply:
That is a good idea and perhaps it needs to be behind a flag if someone finds getting multiple tabs from multiple runs annoying. One issue with most solutions though is they are not robust enough. For example, to open the browser only when the server is successfully up we need to do something like https://stackoverflow.com/questions/32738188/how-can-i-start-the-browser-after-the-server-started-listening/32742904#32742904 but we can't since we don't start the server this way.